src/packages/frontend/editors/slate/slate-react/components/slate.tsx

```diff
@@ -21,12 +21,31 @@ export const Slate = (props: {
 }) => {
   const { editor, children, onChange, value, ...rest } = props;
   const [ticks, setTick] = useState(0);
+
+  // For the editor context we just create it *once* the first
+  // time, then updated it every time by changing the reference.
+  // This avoids an ENORMOUS amount of wasted effort otherwise,
+  // which I found by profiling a large document, and is fine for
+  // our purposes.
   const context: [ReactEditor] = useMemo(() => {
     editor.children = value;
     editor.ticks = ticks;
     Object.assign(editor, rest);
     return [editor];
-  }, [ticks, value, ...Object.values(rest)]);
+  }, []); //[ticks, value, ...Object.values(rest)]);
+  context[0].children = value;
+  context[0].ticks = ticks;
+  Object.assign(context[0], rest);
+
+  // Similarly we use a singleton object for the focused context.
+  // It turns out not doing this with these contexts makes slate
+  // **insanely slow** on large documents.  Doing this... and it
+  // is VERY fast.  It probably took over a month of my life to
+  // understand this, so please don't mess it up again!
```

Comment:
Nice! Congrats on finding this! :rocket:

Reply:
Sadly, it turns out that with more testing, I was wrong :-(.  Back to the grind of virualization/windowing.

(Making the above changes is indeed much faster than nothing... but breaks other things, and is MUCH, much slower than what can be accomplished using windowing.)

Reply:
This is hard.